src/gc.c

```diff
@@ -4069,8 +4073,18 @@ JL_DLLEXPORT jl_value_t *jl_gc_internal_obj_base_ptr(void *p)
         // before the freelist pointer was either live during the last
         // sweep or has been allocated since.
         if (gc_page_data(cell) == gc_page_data(pool->freelist)
-            && (char *)cell < (char *)pool->freelist)
+            && (char *)cell < (char *)pool->freelist) {
             goto valid_object;
+        }
+        else {
+            jl_taggedvalue_t *v = pool->freelist;
+            while (v != NULL) {
+                if (v == cell) {
+                    return NULL;
+                }
+                v = v->next;
```

Comment:
We are seeing some major regression in GC performance with Julia master and wonder if it may be caused by this fix. Presumably if the freelist is long, and we do a lot of `jl_gc_internal_obj_base_ptr` calls, then this would indeed be slow-ish.

Reply:
See also https://github.com/oscar-system/Oscar.jl/issues/2543#issuecomment-1641555936

Reply:
There is likely some low-hanging fruit here.

We may possibly restrict this loop for the subset of the free-list which is contained in the page.

Will open a PR with that.

Reply:
Just for future reference, the PR @d-netto opened was #50599 and has already been merged